File: go/gateway/attestation.go
package gateway

import (
	"github.com/hassard0/vcp-servers/go/sdk"
)

// Attestation is a Provider-signed record of an execution (spec §9,
// schemas/attestation.schema.json). The provider_signature is computed over the
// attestation with its signature block removed (spec §3 rule 4).
type Attestation struct {
	CapabilityID         string         `json:"capability_id"`
	ArgumentHash         string         `json:"argument_hash"`
	ResultHash           string         `json:"result_hash"`
	IdempotencyKey       string         `json:"idempotency_key,omitempty"`
	EffectCommitted      bool           `json:"effect_committed"`
	ObservedExternalRefs []string       `json:"observed_external_refs,omitempty"`
	ProviderSignature    *sdk.Signature `json:"provider_signature,omitempty"`
}

// ResultEnvelope is the Provider -> Gateway result + attestation pair (spec §9).
type ResultEnvelope struct {
	Result      any         `json:"result"`
	Attestation Attestation `json:"attestation"`
}

// Attestation verification reason codes.
const (
	AttestationReasonOK             = "OK"
	AttestationReasonBadSignature   = "ATTESTATION_SIGNATURE_INVALID"
	AttestationReasonCapMismatch    = "ATTESTATION_CAPABILITY_MISMATCH"
	AttestationReasonArgMismatch    = "ATTESTATION_ARGUMENT_MISMATCH"
	AttestationReasonResultMismatch = "ATTESTATION_RESULT_HASH_MISMATCH"
)

// AttestationVerdict is the result of VerifyAttestation.
type AttestationVerdict struct {
	OK         bool
	ReasonCode string
}

// VerifyAttestation verifies a result envelope against what the Gateway authorized
// (spec §9). It MUST confirm, before returning the result to the Planner:
//
//  1. result_hash == sha256(JCS(result)) — the attested result matches the bytes.
//  2. capability_id == the authorized capability.
//  3. argument_hash == the authorized argument hash.
//  4. provider_signature verifies over the attestation without its signature block.
//
// Any failure discards the result (spec §19) by returning OK=false.
func VerifyAttestation(env ResultEnvelope, expectedCapabilityID, expectedArgumentHash string, verifier sdk.Verifier) AttestationVerdict {
	att := env.Attestation

	// 1. Recompute the result hash and compare.
	resultHash, err := sdk.HashJCS(env.Result)
	if err != nil {
		return AttestationVerdict{ReasonCode: AttestationReasonResultMismatch}
	}
	if !constantTimeStringEqual(resultHash, att.ResultHash) {
		return AttestationVerdict{ReasonCode: AttestationReasonResultMismatch}
	}

	// 2 & 3. Bindings to what was authorized (exact identifier comparison).
	if !constantTimeStringEqual(att.CapabilityID, expectedCapabilityID) {
		return AttestationVerdict{ReasonCode: AttestationReasonCapMismatch}
	}
	if !constantTimeStringEqual(att.ArgumentHash, expectedArgumentHash) {
		return AttestationVerdict{ReasonCode: AttestationReasonArgMismatch}
	}

	// 4. Provider signature over the attestation minus its signature block.
	if att.ProviderSignature == nil {
		return AttestationVerdict{ReasonCode: AttestationReasonBadSignature}
	}
	unsigned, err := attestationWithoutSignature(att)
	if err != nil {
		return AttestationVerdict{ReasonCode: AttestationReasonBadSignature}
	}
	ok, err := sdk.VerifyValue(verifier, unsigned, *att.ProviderSignature)
	if err != nil || !ok {
		return AttestationVerdict{ReasonCode: AttestationReasonBadSignature}
	}

	return AttestationVerdict{OK: true, ReasonCode: AttestationReasonOK}
}

// SignAttestation fills result_hash and signs the attestation (Provider side).
func SignAttestation(env *ResultEnvelope, s sdk.Signer) error {
	rh, err := sdk.HashJCS(env.Result)
	if err != nil {
		return err
	}
	env.Attestation.ResultHash = rh
	unsigned, err := attestationWithoutSignature(env.Attestation)
	if err != nil {
		return err
	}
	sig, err := sdk.SignValue(s, unsigned)
	if err != nil {
		return err
	}
	env.Attestation.ProviderSignature = &sig
	return nil
}

func attestationWithoutSignature(att Attestation) (map[string]any, error) {
	att.ProviderSignature = nil
	mp, err := decodeToMap(att)
	if err != nil {
		return nil, err
	}
	delete(mp, "provider_signature")
	return mp, nil
}

File: rust/crates/vcp-gateway/src/attestation.rs
//! Result attestation (§9). Every result carries a Provider-signed attestation.
//! The Gateway MUST verify the signature and that `capability_id` and
//! `argument_hash` match what it authorized, and that `result_hash` matches the
//! returned result, before releasing the (tainted) result to the Planner.
//! Verification failure discards the result (§19).

use serde::{Deserialize, Serialize};
use serde_json::Value;

use vcp_sdk::jcs;
use vcp_sdk::signer::{Signer, Verifier};

/// In-band provider signature.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProviderSignature {
    pub alg: String,
    pub value: String,
}

/// The attestation body (§9).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Attestation {
    pub capability_id: String,
    pub argument_hash: String,
    pub result_hash: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub idempotency_key: Option<String>,
    pub effect_committed: bool,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub observed_external_refs: Option<Vec<String>>,
    pub provider_signature: ProviderSignature,
}

impl Attestation {
    /// JCS bytes the provider signs: the attestation without its
    /// `provider_signature` block.
    pub fn signing_bytes(&self) -> String {
        let mut v = serde_json::to_value(self).expect("attestation serializes");
        if let Value::Object(ref mut map) = v {
            map.remove("provider_signature");
        }
        jcs::canonicalize_value(&v)
    }
}

/// Result + attestation envelope (§9).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AttestedResult {
    pub result: Value,
    pub attestation: Attestation,
}

/// Build a signed attested result (provider side).
pub fn sign_result(
    capability_id: &str,
    argument_hash: &str,
    result: Value,
    effect_committed: bool,
    idempotency_key: Option<String>,
    observed_external_refs: Option<Vec<String>>,
    provider: &dyn Signer,
) -> AttestedResult {
    let result_hash = jcs::hash_value(&result);
    let mut att = Attestation {
        capability_id: capability_id.to_string(),
        argument_hash: argument_hash.to_string(),
        result_hash,
        idempotency_key,
        effect_committed,
        observed_external_refs,
        provider_signature: ProviderSignature {
            alg: provider.alg().to_string(),
            value: String::new(),
        },
    };
    let value = provider.sign(att.signing_bytes().as_bytes());
    att.provider_signature.value = value;
    AttestedResult {
        result,
        attestation: att,
    }
}

/// Why an attestation was rejected (§9).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AttestationError {
    BadSignature,
    CapabilityMismatch,
    ArgumentHashMismatch,
    ResultHashMismatch,
}

/// Verify an attested result against what the Gateway authorized (§9). On any
/// failure the result MUST be discarded.
pub fn verify_attestation(
    attested: &AttestedResult,
    expected_capability_id: &str,
    expected_argument_hash: &str,
    provider_verifier: &dyn Verifier,
) -> Result<(), AttestationError> {
    let att = &attested.attestation;

    // Signature over the attestation body.
    if !provider_verifier.verify(att.signing_bytes().as_bytes(), &att.provider_signature.value) {
        return Err(AttestationError::BadSignature);
    }
    // capability_id and argument_hash must match what was authorized (§9).
    if att.capability_id != expected_capability_id {
        return Err(AttestationError::CapabilityMismatch);
    }
    if att.argument_hash != expected_argument_hash {
        return Err(AttestationError::ArgumentHashMismatch);
    }
    // result_hash must match the actual returned result.
    if att.result_hash != jcs::hash_value(&attested.result) {
        return Err(AttestationError::ResultHashMismatch);
    }
    Ok(())
}
